File: java/exercises/src/test/java/org/artrev/workshop/fp/part1/exercise6/Exercise6Test.java
package org.artrev.workshop.fp.part1.exercise6;

import org.artrev.workshop.fp.tools.Function1;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class Exercise6Test {
    @Test
    public void should_compose_function_with_consumer() {
        // given:
        final Consumer1<Integer> consumer =
                integer -> System.out.println(String.format("consumed: %d", integer));

        final Function1<String, Integer> function =
                Integer::valueOf;

        // when:
        final Consumer1<String> tested =
                Exercise6.composeConsumer(function, consumer);

        // then:
        tested.accept("42");
    }

    @Test
    public void should_compose_function_with_supplier() {
        // given:
        final Supplier<String> supplier = () -> "42";
        final Function1<String, Integer> function =
                Integer::valueOf;

        // when:
        final Supplier<Integer> tested =
                Exercise6.composeSupplier(supplier, function);

        // then:
        assertEquals(new Integer(42), tested.get());
    }

    @Test
    public void should_apply_first_argument() {
        // given:
        final Supplier<String> supplier = () -> "foo";
        final Function1<String, Function1<Integer, String>> function =
                string -> integer -> String.format("%s : %d", string, integer);

        // when:
        final Function1<Integer, String> tested =
                Exercise6.applyFirst(function, supplier);

        // then:
        assertEquals(function.apply("foo").apply(42), tested.apply(42));
    }

    @Test
    public void should_apply_second_argument() {
        // given:
        final Supplier<Integer> supplier = () -> 42;
        final Function1<String, Function1<Integer, String>> function =
                string -> integer -> String.format("%s : %d", string, integer);

        // when:
        final Function1<String, String> tested =
                Exercise6.applySecond(function, supplier);

        // then
        assertEquals(function.apply("foo").apply(42), tested.apply("foo"));
    }
}

File: kotlin/solutions/src/test/kotlin/org/artrev/workshop/fp/part1/exercise6/Exercise6Test.kt
@file:Suppress("PackageName")

package org.artrev.workshop.fp.part1.exercise6

import org.junit.Assert.assertEquals
import org.junit.Test

class Exercise6Test {
    @Test
    fun `should compose function with consumer`() {
        // given:
        val consumer: (Int) -> Unit = { integer ->
            println(String.format("consumed: %d", integer))
        }

        val function: (String) -> Int = { Integer.valueOf(it) }

        // when:
        val tested = consumer.compose(function)

        // then:
        tested("42")
    }

    @Test
    fun `should compose function with the supplier`() {
        // given:
        val supplier = { "42" }

        val function: (String) -> Int = { Integer.valueOf(it) }

        // when:
        val tested = supplier.andThen(function)

        // then:
        assertEquals(42, tested())
    }

    @Test
    fun `should apply the first argument`() {
        // given:
        val supplier = { "foo" }
        val function: (String) -> (Int) -> String = { string ->
            { integer ->
                String.format("%s : %d", string, integer)
            }
        }

        // when:
        val tested = function.applyFirst(supplier)

        // then:
        assertEquals(function("foo")(42), tested(42))
    }

    @Test
    fun `should apply the second argument`() {
        val supplier = { 42 }
        val function: (String) -> (Int) -> String = { string ->
            { integer ->
                String.format("%s : %d", string, integer)
            }
        }

        // when:
        val tested = function.applySecond(supplier)

        // then:
        assertEquals(function("foo")(42), tested("foo"))
    }
}